sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs

```diff
@@ -57,42 +58,93 @@ protected override void Initialize(SonarAnalysisContext context)
 
                     var literal = (LiteralExpressionSyntax)c.Node;
 
-                    var numberWithoutSuffix = ClearNumberTypeSuffix(literal.Token.Text);
-
-                    var decimalParts = numberWithoutSuffix.Split('.');
-                    if (decimalParts.Length > 2)
-                    {
-                        return;
-                    }
-
-                    var hasIrregularPattern = decimalParts
-                        .SelectMany(part => part.Split('_'))
-                        .Select(x => x.Length)
-                        .Skip(1) // skip the first part (1_234 => 234)
-                        .Reverse().Skip(decimalParts.Length == 2 ? 1 : 0) // skip the last if there is a decimal (.234_5 => 234)
-                        .Distinct().Skip(1).Any(); // we expect to have only 1 size of pattern
-
-                    if (hasIrregularPattern)
+                    if (HasIrregularPattern(literal.Token.Text))
                     {
                         c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, literal.GetLocation()));
                     }
                 },
                 SyntaxKind.NumericLiteralExpression);
         }
 
-        private static string ClearNumberTypeSuffix(string numberLiteral)
+        public static bool HasIrregularPattern(string numericToken)
         {
-            if (numberLiteral.EndsWith("UL", StringComparison.OrdinalIgnoreCase))
+            // hexadecimal en binary prefixes
+            var start = numericToken.StartsWith("0x", StringComparison.InvariantCultureIgnoreCase)
+                || numericToken.StartsWith("0b", StringComparison.InvariantCultureIgnoreCase) ? 2 : 0;
+
+            var length = numericToken.Length - start;
+
+            // UL suffix.
+            if (numericToken.EndsWith("UL", StringComparison.OrdinalIgnoreCase))
+            {
+                length -= 2;
+            }
+            // single suffixes
+            else if ("LDFUMldfum".IndexOf(numericToken[numericToken.Length - 1]) != NotFound)
+            {
+                length--;
+            }
```

Comment:
This part does not solve the issue that there will be FP for hexadecimal numbers finishing with `D` or `F`.
Ex:
```
var a = 0xFF_FF_1F; // FP
var a = 0xFF_FF_1D; // FP
```

Reply:
Well, given the discussion I was under the impression that we where not going for groups of 3 for decimals and groups of 2, 4, or 8 for hex and binary. I can live with both, but there were at least some in consistencies (and bugs), that had to be fixed. If there be found agreement on the allowed group sizes, its is easy to add.

Reply:
Yes, I still need to discuss further with another colleague regarding the rule but for now let's assume we don't want a change in the behaviour, as this could be done in a separate PR if needed.

The false positive in the rule I'm talking about is not directly linked to the number of elements of such groups: it is related to the fact that hexedecimal numbers finishing with a `F` or `D` do not behave as expected.

Current behaviour is the following:
```
var a = 0xFF_FF_F1; // Ok - no issue raised, groups of 2
var b = 0xFFF_FFF_FF1; // Ok - no issue raised, groups of 3
var c = 0xFFFF_FFFF_FFF1; // Ok - no issue raised, groups of 4

var d = 0xFF_FF_FF; // KO - issue raised but should not - last group is thought to be of size 1 as the last 'F' of the group is considered as float number suffix - FP
var e = 0xFFF_FFF_FFD; // KO - issue raised but should not - last group is thought to be of size 2 as the last 'D' of the group is considered as double number suffix - FP
var f = 0xFFFF_FFFF_FFFF; // KO - issue raised but should not - FP

var g = 0xFF_FF_FFF; // No issue raised, but it should - grouping is not consistant
var h = 0xFFF_FFF_FFFD; // No issue raised, but it should - grouping is not consistant
var i = 0xFFFF_FFFF_FFFFF; // No issue raised, but it should - grouping is not consistant
```
This is the most important problem with the rule at the moment

Reply:
@christophe-zurn-sonarsource 

The false positives have been fixed, and a minium group size of 2 is also taken into account.